File: ui/lib/backup-state.ts
// The Apps page's backup decisions (design/storage.md §4.4, #298), extracted
// so they execute under `npm test` rather than being asserted in a comment —
// the same rule components/storage/backup-runs.ts follows, for the same
// reason: a tile that reads green over an app nobody has backed up in a month
// is the failure this whole feature exists to prevent.

import type { App, AppBackupState } from './types';

/** What the OVERDUE badge says and shows on hover. null = no badge. */
export interface BackupBadge {
  /** "OVERDUE · 9d" or "OVERDUE · never backed up". */
  label: string;
  /** The api's reason, for the tooltip. */
  title: string;
  /** Where the badge links: the runs table that names which app failed. */
  href: string;
}

/**
 * Compact elapsed time: "9d", "2h", "40m", "12s". Days once past a day,
 * hours once past an hour — a badge has room for one unit.
 */
export function compactAge(ms: number): string {
  const s = Math.max(0, Math.floor(ms / 1000));
  if (s < 60) return `${s}s`;
  const m = Math.floor(s / 60);
  if (m < 60) return `${m}m`;
  const h = Math.floor(m / 60);
  if (h < 24) return `${h}h`;
  return `${Math.floor(h / 24)}d`;
}

/**
 * The badge for an app row. Only an OVERDUE state earns one: §4.4's red is
 * for a backup that did not happen. `never` (a fresh install inside its
 * grace), `unconfigured` (#299's nag — no target, schedule off) and `none`
 * (nothing to back up) render nothing here, and an absent `backup` field —
 * an api that does not derive it — is unknown, never fine, and also
 * renders nothing rather than inventing a state.
 */
export function backupBadge(backup: AppBackupState | undefined, now: number = Date.now()): BackupBadge | null {
  if (!backup || backup.state !== 'overdue') return null;
  const elapsed = backup.lastSuccessAt ? compactAge(now - new Date(backup.lastSuccessAt).getTime()) : 'never backed up';
  return {
    label: `OVERDUE · ${elapsed}`,
    title: backup.reason ? `Backup overdue: ${backup.reason}` : 'Backup overdue',
    href: '/storage',
  };
}

/**
 * The quiet line for the detail drawer: what the state is in words. Every
 * state has one — the drawer is where an operator asking "is this backed
 * up?" gets a full sentence rather than a badge.
 */
export function backupSummary(backup: AppBackupState | undefined, now: number = Date.now()): string {
  if (!backup) return 'Backup state unknown — this api does not report it.';
  switch (backup.state) {
    case 'ok':
      return backup.lastSuccessAt
        ? `Backed up ${compactAge(now - new Date(backup.lastSuccessAt).getTime())} ago.`
        : 'Backed up.';
    case 'overdue':
      return backup.lastSuccessAt
        ? `OVERDUE — last backed up ${compactAge(now - new Date(backup.lastSuccessAt).getTime())} ago.`
        : 'OVERDUE — never backed up.';
    case 'never':
      return 'Not backed up yet — the first scheduled backup is still ahead.';
    case 'unconfigured':
      return 'Backups are not configured on this cluster.';
    case 'none':
      return 'Nothing to back up — no volume of this app is classed critical or state.';
    default:
      return `Backup state: ${backup.state}.`;
  }
}

/**
 * Overdue apps first — §4.4's top billing — then the rest in the order the api
 * listed them (install order). Stable: two overdue apps keep their relative
 * order, so the list does not shuffle between polls.
 */
export function sortAppsOverdueFirst<T extends Pick<App, 'backup'>>(apps: readonly T[]): T[] {
  const overdue = apps.filter((a) => a.backup?.state === 'overdue');
  const rest = apps.filter((a) => a.backup?.state !== 'overdue');
  return [...overdue, ...rest];
}

File: proto/backupstate.go
package proto

import "time"

// Per-app backup state — design/storage.md §4.4's "failure is loud", as one
// signal with three surfaces (geekdojo/geekdojo-brain#298).
//
// The unit is the APP, not the run. A run is a cluster-wide event and can end
// failed because of one app on one offline node while every other app's data
// landed; what an operator needs to know per app is a single question — "has
// this app's data been captured within its cadence?" — and this is the answer
// to it, derived on the api from the backup ledger and served on every
// /api/apps row. The app tile, the alert path and the job feed all read the
// same derivation, so they cannot disagree.

// AppBackupStateKind is where an app stands against its backup cadence.
type AppBackupStateKind string

const (
	// AppBackupOK — a generation still on the target holds every one of the
	// app's backed-up volumes, captured inside the cadence, and nothing newer
	// has failed for it.
	AppBackupOK AppBackupStateKind = "ok"
	// AppBackupOverdue — §4.4's red state. The app's data has not been
	// captured within its cadence plus grace, or has never been captured and
	// the app has been installed longer than that, or the most recent
	// attempt to capture it FAILED (its node was offline, its agent refused,
	// its upload did not land). A backup that did not happen is a failed
	// backup, never a skipped one — this is the state that says so.
	AppBackupOverdue AppBackupStateKind = "overdue"
	// AppBackupNever — no generation has ever captured the app, and it was
	// installed recently enough that the first scheduled run is still ahead.
	// Not red: a fresh install is not a missed backup. It becomes overdue on
	// its own once the cadence plus grace has elapsed.
	AppBackupNever AppBackupStateKind = "never"
	// AppBackupUnconfigured — backups are not set up on this cluster: no
	// target is claimed, or the schedule is turned off. Deliberately NOT
	// overdue: nothing was due. This is the standing-nag state
	// geekdojo/geekdojo-brain#299 owns, and it must not render as OVERDUE.
	AppBackupUnconfigured AppBackupStateKind = "unconfigured"
	// AppBackupNone — the app has nothing a backup would take: every volume
	// its tile declares is `cache` or `bulk`, or it has no tile (a custom
	// compose app) and so no classification. Never overdue. Reason says
	// which.
	AppBackupNone AppBackupStateKind = "none"
)

// AppBackupState is the per-app answer, served as the `backup` field of an
// /api/apps row. Identifiers, times and prose only — nothing key-shaped can
// appear here, and nothing does.
type AppBackupState struct {
	State AppBackupStateKind `json:"state"`
	// Class is the highest §4.2 class among the app's backed-up volumes —
	// `critical` if any volume is, else `state`. It is what decides the
	// alert's severity: a stale password vault is crit, stale app state is
	// warn. Empty for AppBackupNone.
	Class string `json:"class,omitempty"`
	// LastSuccessAt is when a generation last captured EVERY one of this
	// app's backed-up volumes. Absent means never.
	LastSuccessAt *time.Time `json:"lastSuccessAt,omitempty"`
	// LastAttemptAt is when a run last tried — the newest terminal run that
	// reached this app, or that failed before reaching anything.
	LastAttemptAt *time.Time `json:"lastAttemptAt,omitempty"`
	// OverdueSince is when the state became overdue: the moment the cadence
	// plus grace ran out, or the moment the failed attempt ended. Only set
	// when State is overdue; the alert's Since.
	OverdueSince *time.Time `json:"overdueSince,omitempty"`
	// Reason is the sentence beside the state — for overdue, the fan-out's
	// own per-volume reason (off-node, refused, upload did not land) or
	// the age; for the other states, why the app is in it.
	Reason string `json:"reason,omitempty"`
	// Cadence is the schedule the state was judged against, as a duration
	// string, so a tile can say "weekly" beside "9d".
	Cadence string `json:"cadence,omitempty"`
}

// Overdue reports whether the state is §4.4's red one.
func (s AppBackupState) Overdue() bool { return s.State == AppBackupOverdue }

// AppBackupStatus is one app's state with the app named, for the consumers
// that walk every app rather than one row — the alerts aggregator.
type AppBackupStatus struct {
	AppID   string `json:"appId"`
	AppName string `json:"appName"`
	AppBackupState
}
